src/cart/src/gurt/hash.c

```diff
@@ -867,218 +1023,240 @@ link2rlink(d_list_t *link)
-	hlink->hl_key = ((dht->ch_cookie++) << D_HTYPE_BITS) | type;
-}
-
-static int
-hh_op_key_get(struct d_hash_table *hhtab, d_list_t *rlink, void **key_pp)
-{
-	struct d_hlink	*hlink = hh_link2ptr(rlink);
-
-	*key_pp = (void *)&hlink->hl_key;
-	return sizeof(hlink->hl_key);
+	hhash = container_of(htable, struct d_hhash, ch_htable);
+	hlink->hl_key = ((hhash->ch_cookie++) << D_HTYPE_BITS)
+			| (type & D_HTYPE_MASK);
 }
 
 static uint32_t
-hh_op_key_hash(struct d_hash_table *hhtab, const void *key, unsigned int ksize)
+hh_op_key_hash(struct d_hash_table *htable, const void *key, unsigned int ksize)
 {
 	D_ASSERT(ksize == sizeof(uint64_t));
 
-	return (unsigned int)(*(const uint64_t *)key >> D_HTYPE_BITS);
+	return (uint32_t)(*(const uint64_t *)key >> D_HTYPE_BITS);
 }
 
 static bool
-hh_op_key_cmp(struct d_hash_table *hhtab, d_list_t *link,
+hh_op_key_cmp(struct d_hash_table *htable, d_list_t *link,
 	      const void *key, unsigned int ksize)
 {
-	struct d_hlink	*hlink = hh_link2ptr(link);
+	struct d_hlink	*hlink = link2hlink(link);
 
 	D_ASSERT(ksize == sizeof(uint64_t));
+
 	return hlink->hl_key == *(uint64_t *)key;
 }
 
+static uint32_t
+hh_op_rec_hash(struct d_hash_table *htable, d_list_t *link)
+{
+	struct d_hlink	*hlink = link2hlink(link);
+
+	return (uint32_t)(hlink->hl_key >> D_HTYPE_BITS);
+}
+
 static void
-hh_op_rec_addref(struct d_hash_table *hhtab, d_list_t *link)
+hh_op_rec_addref(struct d_hash_table *htable, d_list_t *link)
 {
-	rlink_op_addref(link2rlink(link));
+	rl_op_addref(link2rlink(link));
 }
 
 static bool
-hh_op_rec_decref(struct d_hash_table *hhtab, d_list_t *link)
+hh_op_rec_decref(struct d_hash_table *htable, d_list_t *link)
 {
-	return rlink_op_decref(link2rlink(link));
+	return rl_op_decref(link2rlink(link));
 }
 
 static void
-hh_op_rec_free(struct d_hash_table *hhtab, d_list_t *link)
+hh_op_rec_free(struct d_hash_table *htable, d_list_t *link)
 {
-	struct d_hlink	*hlink = hh_link2ptr(link);
+	struct d_hlink	*hlink = link2hlink(link);
 
 	if (hlink->hl_ops != NULL && hlink->hl_ops->hop_free != NULL)
 		hlink->hl_ops->hop_free(hlink);
 }
 
 static d_hash_table_ops_t hh_ops = {
 	.hop_key_init		= hh_op_key_init,
-	.hop_key_get		= hh_op_key_get,
 	.hop_key_hash		= hh_op_key_hash,
 	.hop_key_cmp		= hh_op_key_cmp,
+	.hop_rec_hash		= hh_op_rec_hash,
 	.hop_rec_addref		= hh_op_rec_addref,
 	.hop_rec_decref		= hh_op_rec_decref,
 	.hop_rec_free		= hh_op_rec_free,
 };
 
 int
-d_hhash_create(uint32_t feats, unsigned int bits, struct d_hhash **htable_pp)
+d_hhash_create(uint32_t feats, uint32_t bits, struct d_hhash **hhash_pp)
 {
-	struct d_hhash	*hhtab;
-	int		 rc;
+	struct d_hhash	*hhash;
+	int		 rc = 0;
 
-	D_ALLOC_PTR(hhtab);
-	if (hhtab == NULL)
-		return -DER_NOMEM;
+	D_ALLOC_PTR(hhash);
+	if (hhash == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
 
 	rc = d_hash_table_create_inplace(feats, bits, NULL, &hh_ops,
-					 &hhtab->ch_htable);
-	if (rc != 0)
-		goto failed;
+					 &hhash->ch_htable);
+	if (rc) {
+		D_FREE_PTR(hhash);
```

Comment:
D_FREE(), and it'll set hhash to NULL for you as well.

Reply:
ok.